tools/diagnostics/nsdiff

```diff
@@ -0,0 +1,875 @@
+    # directories.
+    def print_zipfile(self, filename):
+        f = zipfile.ZipFile(filename)
+        for info in f.infolist():
+            logger.debug(info.filename)
+    
+    # Print the contents of a directory
+    def print_directory(self, dir):
+        onlyfiles = [f for f in listdir(dir) if isfile(join(dir, f))]
+        onlydirs = [f for f in listdir(dir) if isdir(join(dir, f))]
+        logger.debug("In {}:".format(dir))
+        for file in onlyfiles:
+            logger.debug("\t{}".format(file))
+        for dirent in onlydirs:
+            self.print_directory(join(dir, dirent))
+
+    # Extract a tarball
+    def _extract_tarball(self, file):
+        tempdir = None
+        logger.debug("Extracting tarball {}".format(file))
+        if self.log_level <= logging.INFO:
+            self.print_tarfile(file)
+        tempdir = tempfile.mkdtemp()
+        if ( tempdir ):
+            with tarfile.open(file, "r") as tar:
+                tar.extractall(tempdir)
+        else:
+            s = "Failed to create tempdir in which to unpack {}"
+            raise ArchiveError(s.format(file))
+        return tempdir
+
+    # Extract a zipfile
+    def _extract_zipfile(self, file):
+        tempdir = None
+        logger.debug("Extracting zipfile {}".format(file))
+        if self.log_level <= logging.INFO:
+            self.print_zipfile(file)
+        tempdir = tempfile.mkdtemp()
+        if ( tempdir ):
+            with zipfile.ZipFile(file, 'r') as zip:
+                zip.extractall(tempdir)
+        else:
+            s = "Failed to create tempdir in which to unpack {}"
+            raise ArchiveError(s.format(file))
+        return tempdir
+    
+    # Extract a tarball/zipfile into a temprary directory
+    def _extract_archive(self, file):
+        logger.debug("Extracting archive {}".format(file))
+        # Is the file a tarfile (tar, gz2, zip, tar.gz, etc.) or zipfile?
+        if tarfile.is_tarfile(file):
+            return self._extract_tarball(file)
+        elif zipfile.is_zipfile(file):
+            return self._extract_zipfile(file)
+        else:
+            raise ArchiveError("{} must be a tarfile or a zipfile".format(file))
+    
+    def split_name(self, file):
+        abs_file = abspath(file)
+        baseName = basename(abs_file)
+        baseNameElements = baseName.split('.')
+        fileName = None
+        fileExtension = None
+        if len(baseNameElements) > 0:
+            fileName = baseNameElements[0]
+        if len(baseNameElements) > 1:
+            fileExtension = ".".join(baseNameElements[1:])
+        return ( dirname(abs_file), baseName, fileName, fileExtension ) 
+
+    def _mkdir_p(self, path):
+        try:
+            os.makedirs(path)
+        except OSError as exc:  # Python >2.5
+            if exc.errno == errno.EEXIST and os.path.isdir(path):
+                pass
+            else:
+                raise
+    
+    def transform_and_scrub(self, sname, sin, sout):
```

Comment:
1) Please split the method into a number of smaller ones
2) Please avoid too may hierarchical indentation (if-else-if-else-try-if-else....)

Reply:
We can put some TODO comment to do this. But I don't think we should hold up the PR till this is done.
Alex, I agree with the suggestions but are you ok if we put it on our TODO list?